src/braket/default_simulator/openqasm/program_context.py

```diff
@@ -571,6 +745,68 @@ def add_builtin_gate(
                 ctrl_modifiers += [ctrl_mod_ix] * mod.argument.value
             if mod.modifier == GateModifierName.pow:
                 power *= mod.argument.value
+        self.add_gate_instruction(
+            gate_name, target, params, ctrl_modifiers=ctrl_modifiers, power=power
+        )
+
+    @abstractmethod
+    def add_gate_instruction(
+        self, gate_name: str, target: Tuple[int], params, ctrl_modifiers: List[int], power: int
+    ):
+        """Abstract method to add Braket gate to the circuit.
+        Args:
+            gate_name (str): name of the built-in Braket gate.
+            target (Tuple[int]): control_qubits + target_qubits.
+            ctrl_modifiers (List[int]): Quantum state on which to control the
+                operation. Must be a binary sequence of same length as number of qubits in
+                `control-qubits` in target. For example "0101", [0, 1, 0, 1], 5 all represent
+                controlling on qubits 0 and 2 being in the \\|0⟩ state and qubits 1 and 3 being
+                in the \\|1⟩ state.
+            power(float): Integer or fractional power to raise the gate to.
+        """
+
+    @abstractmethod
+    def add_custom_unitary(
+        self,
+        unitary: np.ndarray,
+        target: Tuple[int],
+    ) -> None:
+        """Abstract method to add a custom Unitary instruction to the circuit
+        Args:
+            unitary (np.ndarray): unitary matrix
+            target (Tuple[int]): control_qubits + target_qubits
+        """
+
+    @abstractmethod
+    def add_noise_instruction(self, noise: KrausOperation):
+        """Abstract method to add a noise instruction to the circuit
+        Args:
+            noise (KrausOperation): The noise operation to be added to the circuit
+                - `target` (Tuple[int]): The target qubit or qubits to which the noise operation is applied.
+                - `gamma` (float): The probability of the noise operation occurring. (Optional)
+                - `probability` (float): The probability associated with possible outcome
+                                        of the noise operation. (Optional)
+                - `probabilities` (List[float]): The probabilities associated with each possible outcome
+                                                of the noise operation. (Optional)
+        """
+
+
+class ProgramContext(AbstractProgramContext):
+    def __init__(self):
+        super().__init__()
+        self.circuit = Circuit()
```

Comment:
if this specific program context class has a `self.circuit` attribute, its init should also accept a circuit argument of the same type

Reply:
Quick fix: #183